mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -164,21 +173,26 @@ public void delete(Mailbox mailbox, MailboxMessage message) {
                 .collect(Guavate.toImmutableList());
     }
 
-    private Stream<SimpleMailboxMessage> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
-        Stream<Pair<CassandraMessageDAO.MessageWithoutAttachment, Stream<CassandraMessageDAO.MessageAttachmentRepresentation>>>
-            messageRepresentions = messageDAO.retrieveMessages(messageIds, fetchType, limit).join();
+    private CompletableFuture<Stream<SimpleMailboxMessage>> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
+        CompletableFuture<Stream<Pair<CassandraMessageDAO.MessageWithoutAttachment, Stream<CassandraMessageDAO.MessageAttachmentRepresentation>>>>
+            messageRepresentations = messageDAO.retrieveMessages(messageIds, fetchType, limit);
         if (fetchType == FetchType.Body || fetchType == FetchType.Full) {
-            return CompletableFutureUtil.allOf(messageRepresentions
-                .map((Pair<CassandraMessageDAO.MessageWithoutAttachment, Stream<CassandraMessageDAO.MessageAttachmentRepresentation>> pair) -> 
+            return FluentFutureStream.of(messageRepresentations)
+                .thenComposeOnAll(pair ->
                     attachmentLoader.getAttachments(pair.getRight().collect(Guavate.toImmutableList()))
-                        .thenApply(attachments -> Pair.of(pair.getLeft(), attachments))))
-                .join()
-                .map(Throwing.function(pair -> pair.getLeft()
-                    .toMailboxMessage(pair.getRight()
-                        .stream()
-                        .collect(Guavate.toImmutableList()))));
+                        .thenApply(attachments -> Pair.of(pair.getLeft(), attachments))
+                )
+                .map(pair ->
+                    pair.getLeft()
+                        .toMailboxMessage(pair.getRight()
+                            .stream()
+                            .collect(Guavate.toImmutableList())))
+                .completableFuture();
         } else {
-            return messageRepresentions.map(pair -> pair.getLeft().toMailboxMessage(ImmutableList.of()));
+            return CompletableFutureUtil.map(messageRepresentations, pair ->
```

Comment:
Like I say before I only need one `method` of CompletableFuturUtil here. I do not need to chain method from it, so using the FluentFutureStream will make the code bigger (extra of and completableFuture)